src/parse.rs

```diff
@@ -488,30 +403,31 @@ mod tests {
     fn parse_capability_invalid_test() {
         let (mut send, recv) = mpsc::channel();
         let lines = b"* JUNK IMAP4rev1 STARTTLS AUTH=GSSAPI LOGINDISABLED\r\n";
-        parse_capabilities(lines.to_vec(), &mut send).unwrap();
+        Capabilities::parse(lines.to_vec(), &mut send).unwrap();
         assert!(recv.try_recv().is_err());
     }
 
     #[test]
     fn parse_names_test() {
         let lines = b"* LIST (\\HasNoChildren) \".\" \"INBOX\"\r\n";
         let (mut send, recv) = mpsc::channel();
-        let names = parse_names(lines.to_vec(), &mut send).unwrap();
+        let names = Names::parse(lines.to_vec(), &mut send).unwrap();
         assert!(recv.try_recv().is_err());
         assert_eq!(names.len(), 1);
+        let first = names.iter().next().unwrap();
         assert_eq!(
-            names[0].attributes(),
+            first.attributes(),
```

Comment:
I think it'd be hugely valuable to continue to support indexing here if at all possible. I suspect all that'd be needed is to implement `Index` for `Names` and `Fetches`.

Reply:
I struggled to implement `Index` or `Deref` for this. The issue is the lifetime param on the `Output` associated type, since `type Output = Name<'a>`. The compiler wants the same lifetime on `Output`, but [generic associated types are unstable](https://github.com/rust-lang/rust/issues/44265).

We could probably get around this by replacing `Names` and `Fetches` with a generic container that takes a type param (we could call it.. `ZeroCopy`?), and then implement `Deref` or `Index` the same way that it is done for `ZeroCopy`, since the lifetime would be encapsulated in the generic type. Then `Names` is just `ZeroCopy<Name<'_>>`. When I try this I end up struggling against the constructor function signatures that ouroboros wants in the `TryBuilder` structure to initialize the self ref structure.

So instead I implemented a `get()` function, which trivially maps the underlying vector `get()` function and lets you just get an indexed member from the container. I personally kind of prefer this anyway, since `Index` can panic, but I get that the `Index` trait gives nice `[]` syntax.

If we want to impl `Index` or `Deref` (I think `Deref` would be better?), I am not sure how to make this work in the ouroboros context. I imagine that if it is doable then it is probably kind of simple, but I couldn't get it going.

Reply:
Oof, yeah, that's awkward. You can `impl<'a> Index<usize> for &'a Fetches` ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7e4293bf8789913779e2ab9264a52ce4)), but it means `&fetches[0]` won't work, only `(&fetches)[0]`, which is both hard to discover and ugly. I'm okay with `fn get` :+1: